app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/MapViewModel.kt

```diff
@@ -103,18 +127,59 @@ constructor(val locationSource: LocationSource, private val beaconRepository: Be
    * @author Menzo Bouaissi
    * @since 4.0
    */
-  fun getRandomSong(id: String) {
+  fun getRandomSongAndAddToProfile(beaconId: String) {
     viewModelScope.launch {
-      beaconRepository.getItem(id).collect { fetchedBeacon ->
+      beaconRepository.getItem(beaconId).collect { fetchedBeacon ->
         fetchedBeacon.onSuccess { beacon ->
-          if (beacon.profileAndTrack.isEmpty()) {
-            Log.e("No songs found", "No songs found for the given id")
-            return@onSuccess
+          if (beacon.profileAndTrack.isNotEmpty()) {
+            _retrievedSongs.value = beacon.profileAndTrack.random()
+            trackRepository.addItemsIfNotExist(listOf(_retrievedSongs.value.track))
           }
-          _retrievedSongs.value = beacon.profileAndTrack.random()
         }
         fetchedBeacon.onFailure { exception ->
-          Log.e("No beacons found", "No beacons found for the given id")
+          Log.e("No beacons found", "No beacons found for the given id $exception")
+        }
+      }
+    }
+  }
+
+  fun retrieveSongFromProfileAndAddToBeacon(beaconId: String) {
+
+    viewModelScope.launch {
+      _profileUI.value.profile?.let {
+        profileRepository.getItem(it.firebaseUid).collect { fetchedProfile ->
+          fetchedProfile.onSuccess { profile ->
+            if (profile.topSongs.isNotEmpty()) {
+              addTrackToBeacon(beaconId, profile.topSongs.random()) {}
+            }
+          }
+          fetchedProfile.onFailure { exception ->
+            Log.e("Profile not found", "Profile not found for the given id $exception")
+          }
+        }
+      }
+    }
+  }
+
+  fun addTrackToBeacon(beaconId: String, track: Track, onComplete: (Boolean) -> Unit) {
```

Comment:
Okay, this function is also almost identically there in the BeaconViewModel. Ideally, we'd want this logic to happen in the BeaconConnection, no?
But I understand if you feel that this refactoring is out of scope for your PR :)

Reply:
I put the function in a new file in the utils folder